specification/mysql/resource-manager/Microsoft.DBforMySQL/preview/2022-06-01-privatepreview/mysql.json

```diff
@@ -0,0 +1,2964 @@
+      "properties": {
+        "principalId": {
+          "description": "Principal Id of user assigned identity",
+          "type": "string",
+          "x-ms-client-name": "PrincipalId",
+          "readOnly": true
+        },
+        "clientId": {
+          "description": "Client Id of user assigned identity",
+          "type": "string",
+          "x-ms-client-name": "ClientId",
+          "readOnly": true
+        }
+      },
+      "description": "Metadata of user assigned identity."
+    },
+    "CutoverMigrationRequest": {
+      "type": "object",
+      "description": "Cutover migration request",
+      "properties": {
+        "dnsChangeMode": {
+          "type": "string",
+          "enum": [
+            "Update",
+            "Revert",
+            "None"
+          ],
+          "x-ms-enum": {
+            "name": "DnsChangeMode",
+            "modelAsString": true
+          },
+          "description": "The DNS Mode"
+        }
+      }
+    },
+    "MigrationAssessmentRequest": {
+      "type": "object",
+      "properties": {
+        "sourceServerResourceId": {
+          "type": "string",
+          "description": "Source Server Resource Id to verify and perform migration assessment on."
+        }
+      },
+      "description": "Request from client to perform migration assessment"
+    },
+    "MigrationAssessmentResource": {
+      "type": "object",
+      "properties": {
+        "sourceServerResourceId": {
+          "type": "string",
+          "description": "Source Server Resource Id to verify and perform migration assessment on."
+        },
+        "targetPlatformProviderId": {
+          "type": "string",
+          "description": "Target platform(Flexible Server) provider ID"
+        },
+        "targetServerDetails": {
+          "$ref": "#/definitions/TargetServerDetails"
+        },
+        "summary": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/AssessmentSummaryItem"
+          },
+          "readOnly": true,
+          "x-ms-identifiers": [
+            "featureName"
+          ],
+          "description": "Represents a summary of Migration Assessment"
+        },
+        "replicaAssessmentReports": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/MigrationAssessmentResource"
+          },
+          "readOnly": true,
+          "x-ms-identifiers": [
+            "sourceServerResourceId"
+          ],
+          "description": "Represents Migration assessment resource of replica servers"
+        },
+        "assessmentReportCompletionDateTime": {
+          "type": "string",
+          "format": "date-time",
+          "readOnly": true,
+          "description": "Assessment report completion date time"
+        }
+      },
+      "description": "Represents a resource name availability."
+    },
+    "TargetServerDetails": {
+      "type": "object",
+      "properties": {
+        "serverResource": {
```

Comment:
>serverResource [](http://example.com/codeflow?start=9&length=14)

If this is in ARM resourceId format, I'd recommend serverResourceId

Reply:
The ServerResource will hold all the properties of a server
The object will look like the one in this file : specification\mysql\resource-manager\Microsoft.DBforMySQL\preview\2022-06-01-privatepreview\examples\ServerGet.json

What is a arm resource Id model?

Reply:
Right, missed that. Actually, the RPC doesn't allow embedding a resource in another resource. I'm trying to get a reading on whether that also applies to POST responses. In any case, this object is already called targetServerDetails, so I think "resource" would actually be a clearer name for this property.

Reply:
The ARM guidance is that embedding a resource in a resource is not allowed in the resource model (GET/PUT/PATCH/DELETE). Instead, include a property that contains the resource ID of the other resource. For POST, that guidance is recommend but not required if there is a good reason to embed the resource. It's critical to ensure that there are no leaks across RBAC though. Can you ensure that anyone using this POST API already has RBAC read access to the server resource? If so, and you think it's important to have the full resource in this payload instead of the resource ID of the server, it's OK to leave this as an embedded resource.

Reply:
Here the ServerResource( which is a target server) in response is not yet created. As part of the assessment, we are just recommending a target server with set of configurations based on the source server.

Reply:
Ah, so this is not a resource, but is constructed. Given that, it seems like this should just be the property bag. You don't have anything useful to fill in for systemData, id, tags, etc. right? Would it be better to call it properties and just include the properties?

Reply:
Sure, that's the usage model I visualized. I was suggesting that the serverResource property should contain the entire complex object, but without the ARM resource envelope (top-level properties like name, type, id, location, systemData, tags (we usually call it the property bag, since it's the contents of an object named "properties")). Those top-level properties need to be populated later at create time, not at assessment time. Consequently, if they are in the assessment payload, any values they have are meaningless. Does that make sense?